### ISSUE TITLE
Support Python 3.9

### ISSUE THREAD

**Issue by park-junha** [tensorflow/tensorflow#44485]
<em>Please make sure that this is a feature request. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:feature_template</em>

**System information**
- TensorFlow version (you are using): 2.3.1
- Are you willing to contribute it (Yes/No): No (willing to do trivial patches)

**Describe the feature and the current behavior/state.**
TensorFlow should be supported on Python 3.9.

**Will this change the current api? How?** Significant changes will likely need to be made.

**Who will benefit with this feature?**
Anyone on Python 3.9

**Any Other info.**

**Comment by mihaimaruseac** [comment]
We have not released anything for python3.9.

Release process is as follows:

1. Ensure all of our dependencies support python3.9 (not yet there)
2. Ensure our code works with all these dependencies on all supported versions of python (at the moment, at least `gast` needs additional support work).
3. Release nightly version supporting the new python version
4. Release a full release.

As we are already doing the 2.4 release, python3.9 support for TF will come in 2.5 release at the earliest.

@ravikyram, @amahendrakar, @rmothukuru  let's deduplicate to this issue all py39 requests/issues. I'll monitor this one and provide updates as we start adding support.

**Comment by rmothukuru** [comment]
@mihaimaruseac,
Sure

**Comment by mihaimaruseac** [comment]
> I may be wrong, but I believe this is the case:
> 
> ```
> python3.8 -m pipdeptree -p tensorflow | grep "^  -" | sed "s/  - //" | sed "s/ \[.*]//" | xargs python3.9 -m pip install
> ```
> 
> which is essentially
> 
> ```
> python3.9 -m pip install absl-py astunparse flatbuffers gast google-pasta grpcio h5py keras-preprocessing numpy opt-einsum protobuf six tensorboard tensorflow-estimator termcolor typing-extensions wheel wrapt
> ```
> 
> works fine for me. Am I missing anything?

(from https://github.com/tensorflow/tensorflow/issues/40840#issuecomment-728760033)

The main issue is that we also have C++ code and we need the needed dependencies to also upgrade for their C++ code. By this time, likely most are solved, so 2.5 release will have py3.9 support.

We should get py3.9 in nightly soon.

**Comment by mihaimaruseac** [comment]
So, all of our depdencies have a release for py3.9 but our `setup.py` upper bounds to versions which are below that release. We will need to increase these bounds while making sure we don't break existing usecases/tests. Will slowly happen over December.

**Comment by willemavjc** [comment]
> So, all of our depdencies have a release for py3.9 but our `setup.py` upper bounds to versions which are below that release. We will need to increase these bounds while making sure we don't break existing usecases/tests. Will slowly happen over December.

@mihaimaruseac Any update? (Quite concerned here as we did an upgrade to 3.9 for some various reasons such Python's core enhancements but totally zapped out the TensorFlow's upper bound T_T)

**Comment by alanpurple** [comment]
I've built tf 2.4 for python 3.9.1 on windows 10 and ubuntu 18.04 and they work fine, no problem at all

**Comment by willemavjc** [comment]
> I've built tf 2.4 for python 3.9.1 on windows 10 and ubuntu 18.04 and they work fine, no problem at all

Curious to know how you did that since you cannot use it because of this “stupid” dependency check. Oh you mean you built it from source? If that’s so, would you mind sharing the commands/options you used?

thanks!

**Comment by ahtik** [comment]
@willemavjc I can share my build from source setup. Win 10 and works with python 3.9.1, TF 2.4.
Python 3.9 installed in c:\python39\ dir.
```
# Copy https://github.com/bazelbuild/bazel/releases/download/3.7.2/bazel-3.7.2-windows-x86_64.exe to somewhere in PATH and rename to bazel.exe 
# Make sure you have Visual Studio 2019 Build Tools installed (located under C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC)
# Make sure cudnn-8.0.5.39 is unpacked into C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v11.0 respective dirs.
# Ensure your env vars contain something like this:
# CUDA_PATH=C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v11.0
# CUDA_TOOLKIT_PATH=%CUDA_PATH%
# PATH=%CUDA_PATH%\bin;%PATH%
# PATH=%CUDA_PATH%\extras\CUPTI\libx64;%PATH%

# Using PowerShell 7.1.0
git clone https://github.com/tensorflow/tensorflow.git tf24build
cd tf24build
git checkout v2.4.0
\python39\python -m venv .venv
.\.venv\Scripts\Activate.ps1
pip install -U pip

# Now, depending on how fresh is your Windows installation, you need either numpy==1.19.3 or or 1.19.4.
# See https://github.com/numpy/numpy/issues/16744 for details
pip install six numpy==1.19.4 wheel keras_applications keras_preprocessing --no-deps

$Env:BAZEL_VC = "C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC"
# Adjust the Compute Capabilities to match your target GPU!
$Env:TF_CUDA_COMPUTE_CAPABILITIES = "7.5"
$Env:CC_OPT_FLAGS = "/arch:AVX2"
$Env:TF_NEED_CUDA = "1"

# Check the prompts but should be just enter for everything
.\configure.cmd

bazel build --config=opt --define=no_tensorflow_py_deps=true //tensorflow/tools/pip_package:build_pip_package

# Generates tensorflow_2.4_py39.whl
bazel-bin\tensorflow\tools\pip_package\build_pip_package tensorflow_2.4_py39

# Create a new venv and install the new wheel, test it's health with
# python -c "import tensorflow as tf;print(tf.reduce_sum(tf.random.normal([1000, 1000])));print(tf.config.list_physical_devices('GPU'))"

**Comment by byronyi** [comment]
Sorry but AFAIK protobuf does not support Python 3.9 properly with their Linux wheels. See https://github.com/protocolbuffers/protobuf/issues/7978.

**Comment by ahtik** [comment]
Ah, I think I also left out a few bits here...
Because grpcio that tensorflow depends on is not having the python 3.9 wheels (but newer version has), my setup to install the wheels is twofold:
```
# Requirements from https://github.com/tensorflow/tensorflow/blob/master/tensorflow/tools/pip_package/setup.py
pip install "grpcio==1.34.0" "absl-py~=0.10" "astunparse~=1.6.3" "flatbuffers~=1.12.0" "google_pasta~=0.2" "h5py~=3.1.0" "keras_preprocessing~=1.1.2" "numpy~=1.19.2" "opt_einsum~=3.3.0" "protobuf>=3.9.2" "six~=1.15.0" "termcolor~=1.1.0" "typing_extensions~=3.7.4" "wheel~=0.35" "wrapt~=1.12.1" "gast==0.3.3"

pip install --no-deps C:\tmp\tensorflow-2.4.0-cp39-cp39-win_amd64.whl
```

This makes it easier to also manage the protobuf installation separately: we do use the regular python version and do not seem to have issues with python 3.9. But for Windows there is also an unofficial Python 3.9 protobuf wheel for Windows at https://www.lfd.uci.edu/~gohlke/pythonlibs/#protobuf (not the latest but 3.12.3, but good enough).

**Comment by mihaimaruseac** [comment]
protobuf and gast are current blockers to have 100% tests passing

**Comment by byronyi** [comment]
> protobuf and gast are current blockers to have 100% tests passing

[gast 0.4.0](https://github.com/serge-sans-paille/gast/releases/tag/0.4.0) seems to support Python 3.9. Haven't tested it yet though.

**Comment by alanpurple** [comment]
@willemavjc 
only problem blocked me was grpcio 1.32.0, which I ignored using --no-dependency option


I've built tensorflow from source of course

cuda 11.2
cudnn 8.0.5
branch - r2.4

**Comment by mihaimaruseac** [comment]
> > protobuf and gast are current blockers to have 100% tests passing
> 
> [gast 0.4.0](https://github.com/serge-sans-paille/gast/releases/tag/0.4.0) seems to support Python 3.9. Haven't tested it yet though.

It does, but I think @mdanatg has a change to update and there are a few things that need to change to make all tests pass.

**Comment by mdanatg** [comment]
The change to upgrade to 0.4 is ready and just needs one final approval. We should be able to resolve this shortly after the approver is back in office, on Jan 5.

**Comment by bhack** [comment]
> The change to upgrade to 0.4 is ready and just needs one final approval. We should be able to resolve this shortly after the approver is back in office, on Jan 5.

This kind of info is related, more in general, to a quite old open topic (since 2016) as you can see https://github.com/tensorflow/tensorflow/pull/46039#issuecomment-752087368

And this specific issue was exposed from October (https://github.com/tensorflow/tensorflow/issues/44146) so we had two months to be exposed to concurrent contributions conflicts.

**Comment by bhack** [comment]
https://github.com/protocolbuffers/protobuf/pull/8121

**Comment by raul-parada** [comment]
can you share the generated tensorflow-2.4.0-cp39-cp39-win_amd64.whl file?

**Comment by tomcharnock** [comment]
When building for mac Big Sur it seems that you need to rename the whl from tensorflow-2.4.0-cp39-cp39-macosx_11_0_x86_64.whl to tensorflow-2.4.0-cp39-cp39-macosx_10_16_x86_64.whl, but after installing gast==0.4, hypy==3.1.0 and grpcio==1.34.0 everything is working perfectly (so far ;))

**Comment by GuillemGSubies** [comment]
Any progress here? I would really appreciate being able to use Python 3.9.

**Comment by mihaimaruseac** [comment]
Not yet. We should support py3.9 in nightly in about a few weeks.

**Comment by byronyi** [comment]
> Python 3.9.2 release is scheduled to happen in two weeks according to https://www.python.org/dev/peps/pep-0596
> 
> Perhaps work should start now on testing against [Python 3.10a4](https://www.python.org/downloads/release/python-3100a4) so the community does not face yet another 7-month delay.

Python 3.9 is released at October 5th, 2020. Or are you referring to something else by “7-month delay”?

**Comment by mihaimaruseac** [comment]
TF has a large number of dependencies. We cannot make progress until all dependencies are updated.

Right now protobuf is still lacking a [`py39` wheel](https://pypi.org/project/protobuf/#files). This means TF is unable to make real progress anyway (as we need the `py39` wheel due to the C++ protobuf interface)

**Comment by mihaimaruseac** [comment]
We have to wait until https://github.com/protocolbuffers/protobuf/pull/8121 lands and a new protobuf containing it (or similar) is released.

**Comment by amay428** [comment]
Hi, any updates yet? Has it been released?

**Comment by supercmmetry** [comment]
When will tensorflow support python 3.9?

**Comment by cclauss** [comment]
Not until after there is a [protobuf release](https://github.com/protocolbuffers/protobuf/releases) that contains the protocolbuffers/protobuf#8121 mods.

**Comment by JJ** [comment]
@cclauss are you implying that it will not happen until the _next_ release, since there was a release 5 days ago that did not include it?

**Comment by cclauss** [comment]
5 days ago was a ___pre___-release

**Comment by amay428** [comment]
@cclauss Can you share any links for web pages with more details about the pre-release? I can't seem to find anything. 
Thanks!

**Comment by ahtik** [comment]
@amay428 https://pypi.org/project/protobuf/3.15.0rc1/#files

**Comment by amay428** [comment]
@ahtik I meant TensorFlow, not protobuf, but thanks

**Comment by mihaimaruseac** [comment]
TensorFlow **will** suport Python 3.9 **first** in **nightly** and then in **Tensorflow 2.5**. But this will happen **only** after all dependencies are available.

**Comment by Nizarazo** [comment]
Hi

Any estimation date when TensorFlow will support Python 3.9 in nightly? 
And any estimation of release date of Tensorflow 2.5?
Is it stable to use the nightly until Tensorflow 2.5 is released?

Thanks

**Comment by cclauss** [comment]
`nightly` is for non-production use so `nightly` != `stable` by definition.

**Comment by mihaimaruseac** [comment]
Currently blocked on Windows by #46902 caused by MSVC / Python 3.9 defines (https://www.gitmemory.com/issue/OpenImageIO/oiio/2799/752372181). As long as upstream projects are not working properly we cannot do anything.

**Comment by yoondray** [comment]
I had the exactly same problems, I can’t wait the date that TF support to Python 3.9. 
I have been searched a lot of ways to use TF to 3.9, eventually, Decide to download lower version Python 3.5.8 and Use different Interpreter by Pycharm.

**Comment by cclauss** [comment]
A bit of a drag that [Python 3.9.2 is shipping](https://www.python.org/downloads/release/python-392/) but TF still does not support Py39.  However, If you need to run TF now, I would recommend looking at [Python 3.8](https://www.python.org/downloads/release/python-388/) instead of 3.5 which is already [end of life](https://devguide.python.org/devcycle/#end-of-life-branches).

**Comment by ahtik** [comment]
The [`copysign` issue](https://bugs.python.org/issue42120) got fixed by Python 3.9.1 by [removing it](https://github.com/python/cpython/pull/23326)

https://docs.python.org/3/whatsnew/changelog.html
![image](https://user-images.githubusercontent.com/140952/108675415-0f6cfb80-74e7-11eb-9ad3-e1af8b4f149f.png)

https://github.com/microsoft/vcpkg/pull/16315 is getting this fix in by https://github.com/microsoft/vcpkg/issues/15369, if anyone wants to track one of the remaining upstream deps preventing the Python 3.9 upgrade.

protobuf got the 3.15.1 out with Python 3.9 a few days ago (unfortunately without the Windows build, but that was already the case with Python 3.8).

**Comment by sergiomora03** [comment]
Hi, any updates yet? Has it been released?

**Comment by bhack** [comment]
Waiting for https://github.com/tensorflow/tensorflow/pull/44282

**Comment by AlbusNoir** [comment]
(just a follow comment to get notified when updates happen)

**Comment by 11lll** [comment]
Hi, looking forward to TF on Python 3.9.

**Comment by ras0219-msft** [comment]
microsoft/vcpkg#16315 has been merged

**Comment by mihaimaruseac** [comment]
We have py39 wheels in recent `tf-nightly` and `tf-nightly-cpu` projects. Pending `tf-nightly-gpu`.

MacOS is not yet supported. Only Linux and Windows for now, not all of them are in all of the above projects yet due to other unrelated failures. Please test nightly before the 2.5 release so we can get 2.5 release with good py39 support.

**Comment by ahtik** [comment]
I do get a build failure when trying to build the latest master (1627cd9e368) Windows wheel for Python 3.9.

I wonder if the following error is one of the known issues or worth reporting? CUDA 11.2, compute capabilities 7.5, 8.6.

```
.\tensorflow/core/platform/ram_file_system.h(197): error C2039: 'lstrcatA': is not a member of 'absl'
external/com_google_absl\absl/strings/str_split.h(51): note: see declaration of 'absl'
.\tensorflow/core/platform/ram_file_system.h(197): error C2664: 'LPSTR lstrcatA(LPSTR,LPCSTR)': cannot convert argument 2 from '_Ty1' to 'LPCSTR'
        with
        [
            _Ty1=const std::string
        ]
.\tensorflow/core/platform/ram_file_system.h(197): note: No user-defined-conversion operator available that can perform this conversion, or the operator cannot be called
C:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\um\winbase.h(2794): note: see declaration of 'lstrcatA'
```

**Comment by samanemami** [comment]
> We have py39 wheels in recent `tf-nightly` and `tf-nightly-cpu` projects. Pending `tf-nightly-gpu`.
> 
> MacOS is not yet supported. Only Linux and Windows for now, not all of them are in all of the above projects yet due to other unrelated failures. Please test nightly before the 2.5 release so we can get 2.5 release with good py39 support.

Still does not work for me (windows - py39)!

**Comment by mihaimaruseac** [comment]
@ahtik It seems you cloned exactly when the repo was failing :( Cloning at de4bc85341d4ffb3e0e8a2f40267459b04c1f2a1 should guarantee a passing Windows build (at least on our infra it was passing).

@samanemami: what where you trying to do and what was the error?

All: All windows and Linux wheels for Py3.9 have been updated. If nothing breaks, over the weekend we should get these wheels updated daily.

**Comment by samanemami** [comment]
@mihaimaruseac The problem has solved.
I finally decided to downgrade my python version and installed the TF for py38.
Looking forward to having TF for py39.
Thanks

**Comment by mihaimaruseac** [comment]
@samanemami: tf-nightly works with py39 on Windows now. TF proper needs to wait for the next release as we cannot just patch old releases for new Python versions.

**Comment by webia1** [comment]
How did you install it @mihaimaruseac ? 
 Thanks for feedbacks,..

```shell 
 ~ sw_vers
ProductName:	macOS
ProductVersion:	11.2.2
BuildVersion:	20D80

 ~ python --version
Python 3.9.1
 
~ pip install tensorflow
ERROR: Could not find a version that satisfies the requirement tensorflow
ERROR: No matching distribution found for tensorflow

 ~ pip install tf-nightly
ERROR: Could not find a version that satisfies the requirement tf-nightly
ERROR: No matching distribution found for tf-nightly
```

**Comment by byronyi** [comment]
Use python -m pip install instead of pip when you have multiple installations

**Comment by webia1** [comment]
> python -m pip

Dear @byronyi  Already tried that, the same result:

```shell
 ~ python --version
Python 3.9.1

 ~ python -m pip install tensorflow
ERROR: Could not find a version that satisfies the requirement tensorflow
ERROR: No matching distribution found for tensorflow
```

**Comment by cclauss** [comment]
@webia1 You are on a Mac and [macOS is not yet supported](https://github.com/tensorflow/tensorflow/issues/44485#issuecomment-786086270).

**Comment by webia1** [comment]
> @webia1 You are on a Mac and [macOS is not yet supported](https://github.com/tensorflow/tensorflow/issues/44485#issuecomment-786086270).

:sob:

**Comment by byronyi** [comment]
API compatibility test fails:

```
ERROR:tensorflow:1 differences found between API and golden.
E0301 10:14:53.017659 140661472663360 api_compatibility_test.py:294] 1 differences found between API and golden.
ERROR:tensorflow:    Change detected in python object: tensorflow.types.experimental.
E0301 10:14:53.017725 140661472663360 api_compatibility_test.py:315]     Change detected in python object: tensorflow.types.experimental.
ERROR:tensorflow:    
  path: "tensorflow.types.experimental"
  tf_module {
    member {
      name: "TensorLike"
-     mtype: "typing.Union"
+     mtype: "<class \'typing._UnionGenericAlias\'>"
    }
  }

E0301 10:14:53.017775 140661472663360 api_compatibility_test.py:316]     
  path: "tensorflow.types.experimental"
  tf_module {
    member {
      name: "TensorLike"
-     mtype: "typing.Union"
+     mtype: "<class \'typing._UnionGenericAlias\'>"
    }
  }
```

cc @mihaimaruseac

**Comment by mdanatg** [comment]
The API compatibility tests are brittle around `typing` objects, especially across versions. We have some normalization for Python 3.7, and likely need to repeat it with 3.9. If that's the sole failure you see, you can safely ignore it until it's patched.

**Comment by Ierezell** [comment]
Hi all, just an update FYI. 

I just loaded a TFBertModel in huggingface, and runned finetunning with tensorflow. 
- pip 21.0.1
- python 3.9.2
- tf-nightly 2.5.0.dev20210228
- linux kernel zen 5.11.1 (Archlinux)

However, I got errors with cuda but for some light operations on cpu it seems to be working.

**Comment by byronyi** [comment]
> The API compatibility tests are brittle around `typing` objects, especially across versions. We have some normalization for Python 3.7, and likely need to repeat it with 3.9. If that's the sole failure you see, you can safely ignore it until it's patched.

Another error lies in `//tensorflow/python/module:module_test`:

```
================================================================================
==================== Test output for //tensorflow/python/module:module_test:
Running tests under Python 3.9.2: /usr/local/bin/python3.9
[ RUN      ] AbcTest.testAbstract
INFO:tensorflow:time(__main__.AbcTest.testAbstract): 0.0s
I0301 10:18:35.056211 140670264665920 test_util.py:2082] time(__main__.AbcTest.testAbstract): 0.0s
[  FAILED  ] AbcTest.testAbstract
[ RUN      ] AbcTest.testConcrete
INFO:tensorflow:time(__main__.AbcTest.testConcrete): 0.0s
I0301 10:18:35.057143 140670264665920 test_util.py:2082] time(__main__.AbcTest.testConcrete): 0.0s
[       OK ] AbcTest.testConcrete
[ RUN      ] AbcTest.test_session
[  SKIPPED ] AbcTest.test_session
[ RUN      ] FlattenTest.test_attribute_traversal_key
2021-03-01 10:18:35.069777: I tensorflow/core/platform/cpu_feature_guard.cc:142] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX512F FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2021-03-01 10:18:35.079878: W tensorflow/python/util/util.cc:348] Sets are not currently considered sequences, but this may change in the future, so consider avoiding using them.
INFO:tensorflow:time(__main__.FlattenTest.test_attribute_traversal_key): 0.02s
I0301 10:18:35.080415 140670264665920 test_util.py:2082] time(__main__.FlattenTest.test_attribute_traversal_key): 0.02s
[       OK ] FlattenTest.test_attribute_traversal_key
[ RUN      ] FlattenTest.test_attributes_to_ignore
INFO:tensorflow:time(__main__.FlattenTest.test_attributes_to_ignore): 0.03s
I0301 10:18:35.108479 140670264665920 test_util.py:2082] time(__main__.FlattenTest.test_attributes_to_ignore): 0.03s
[       OK ] FlattenTest.test_attributes_to_ignore
[ RUN      ] FlattenTest.test_flatten(<function FlattenTest.<lambda> at 0x7fee25b8aa60>)
INFO:tensorflow:time(__main__.FlattenTest.test_flatten(<function FlattenTest.<lambda> at 0x7fee25b8aa60>)): 0.0s
I0301 10:18:35.109399 140670264665920 test_util.py:2082] time(__main__.FlattenTest.test_flatten(<function FlattenTest.<lambda> at 0x7fee25b8aa60>)): 0.0s
[       OK ] FlattenTest.test_flatten(<function FlattenTest.<lambda> at 0x7fee25b8aa60>)
[ RUN      ] FlattenTest.test_flatten(<class 'list'>)
INFO:tensorflow:time(__main__.FlattenTest.test_flatten(<class 'list'>)): 0.0s
I0301 10:18:35.110031 140670264665920 test_util.py:2082] time(__main__.FlattenTest.test_flatten(<class 'list'>)): 0.0s
[       OK ] FlattenTest.test_flatten(<class 'list'>)
[ RUN      ] FlattenTest.test_flatten(<class 'tuple'>)
INFO:tensorflow:time(__main__.FlattenTest.test_flatten(<class 'tuple'>)): 0.0s
I0301 10:18:35.110611 140670264665920 test_util.py:2082] time(__main__.FlattenTest.test_flatten(<class 'tuple'>)): 0.0s
[       OK ] FlattenTest.test_flatten(<class 'tuple'>)
[ RUN      ] FlattenTest.test_flatten(<function <lambda> at 0x7fee25b8a940>)
INFO:tensorflow:time(__main__.FlattenTest.test_flatten(<function <lambda> at 0x7fee25b8a940>)): 0.0s
I0301 10:18:35.111363 140670264665920 test_util.py:2082] time(__main__.FlattenTest.test_flatten(<function <lambda> at 0x7fee25b8a940>)): 0.0s
[       OK ] FlattenTest.test_flatten(<function <lambda> at 0x7fee25b8a940>)
[ RUN      ] FlattenTest.test_raises_error_with_path
INFO:tensorflow:time(__main__.FlattenTest.test_raises_error_with_path): 0.0s
I0301 10:18:35.112239 140670264665920 test_util.py:2082] time(__main__.FlattenTest.test_raises_error_with_path): 0.0s
[       OK ] FlattenTest.test_raises_error_with_path
[ RUN      ] FlattenTest.test_session
[  SKIPPED ] FlattenTest.test_session
[ RUN      ] FlattenTest.test_with_path
INFO:tensorflow:time(__main__.FlattenTest.test_with_path): 0.0s
I0301 10:18:35.114492 140670264665920 test_util.py:2082] time(__main__.FlattenTest.test_with_path): 0.0s
[       OK ] FlattenTest.test_with_path
[ RUN      ] ForwardMethodsTest.testEntersNameScope_call
2021-03-01 10:18:35.124874: I tensorflow/compiler/mlir/mlir_graph_optimization_pass.cc:176] None of the MLIR Optimization Passes are enabled (registered 2)
2021-03-01 10:18:35.125344: I tensorflow/core/platform/profile_utils/cpu_utils.cc:114] CPU Frequency: 2400000000 Hz
INFO:tensorflow:time(__main__.ForwardMethodsTest.testEntersNameScope_call): 0.02s
I0301 10:18:35.132264 140670264665920 test_util.py:2082] time(__main__.ForwardMethodsTest.testEntersNameScope_call): 0.02s
[       OK ] ForwardMethodsTest.testEntersNameScope_call
[ RUN      ] ForwardMethodsTest.testEntersNameScope_concreteFunction
INFO:tensorflow:time(__main__.ForwardMethodsTest.testEntersNameScope_concreteFunction): 0.01s
I0301 10:18:35.138598 140670264665920 test_util.py:2082] time(__main__.ForwardMethodsTest.testEntersNameScope_concreteFunction): 0.01s
[       OK ] ForwardMethodsTest.testEntersNameScope_concreteFunction
[ RUN      ] ForwardMethodsTest.testFunctionType
INFO:tensorflow:time(__main__.ForwardMethodsTest.testFunctionType): 0.0s
I0301 10:18:35.139103 140670264665920 test_util.py:2082] time(__main__.ForwardMethodsTest.testFunctionType): 0.0s
[       OK ] ForwardMethodsTest.testFunctionType
[ RUN      ] ForwardMethodsTest.test_session
[  SKIPPED ] ForwardMethodsTest.test_session
[ RUN      ] ModuleTrackingTest.test_non_ctor_submodule
INFO:tensorflow:time(__main__.ModuleTrackingTest.test_non_ctor_submodule): 0.0s
I0301 10:18:35.140079 140670264665920 test_util.py:2082] time(__main__.ModuleTrackingTest.test_non_ctor_submodule): 0.0s
[       OK ] ModuleTrackingTest.test_non_ctor_submodule
[ RUN      ] ModuleTrackingTest.test_session
[  SKIPPED ] ModuleTrackingTest.test_session
[ RUN      ] ModuleTrackingTest.test_submodules
INFO:tensorflow:time(__main__.ModuleTrackingTest.test_submodules): 0.0s
I0301 10:18:35.141992 140670264665920 test_util.py:2082] time(__main__.ModuleTrackingTest.test_submodules): 0.0s
[       OK ] ModuleTrackingTest.test_submodules
[ RUN      ] NameScopeTest.test_memoized_in_tf2
INFO:tensorflow:time(__main__.NameScopeTest.test_memoized_in_tf2): 0.0s
I0301 10:18:35.142328 140670264665920 test_util.py:2082] time(__main__.NameScopeTest.test_memoized_in_tf2): 0.0s
[       OK ] NameScopeTest.test_memoized_in_tf2
[ RUN      ] NameScopeTest.test_not_memoized_in_tf1
INFO:tensorflow:time(__main__.NameScopeTest.test_not_memoized_in_tf1): 0.0s
I0301 10:18:35.142561 140670264665920 test_util.py:2082] time(__main__.NameScopeTest.test_not_memoized_in_tf1): 0.0s
[  SKIPPED ] NameScopeTest.test_not_memoized_in_tf1
[ RUN      ] NameScopeTest.test_session
[  SKIPPED ] NameScopeTest.test_session
[ RUN      ] TestModuleNaming.test_construct_in_scope
INFO:tensorflow:time(__main__.TestModuleNaming.test_construct_in_scope): 0.0s
I0301 10:18:35.142951 140670264665920 test_util.py:2082] time(__main__.TestModuleNaming.test_construct_in_scope): 0.0s
[       OK ] TestModuleNaming.test_construct_in_scope
[ RUN      ] TestModuleNaming.test_ctor_error_closes_name_scope
INFO:tensorflow:time(__main__.TestModuleNaming.test_ctor_error_closes_name_scope): 0.0s
I0301 10:18:35.143225 140670264665920 test_util.py:2082] time(__main__.TestModuleNaming.test_ctor_error_closes_name_scope): 0.0s
[       OK ] TestModuleNaming.test_ctor_error_closes_name_scope
[ RUN      ] TestModuleNaming.test_ctor_error_handles_ctor_not_opening_name_scope
INFO:tensorflow:time(__main__.TestModuleNaming.test_ctor_error_handles_ctor_not_opening_name_scope): 0.0s
I0301 10:18:35.143461 140670264665920 test_util.py:2082] time(__main__.TestModuleNaming.test_ctor_error_handles_ctor_not_opening_name_scope): 0.0s
[       OK ] TestModuleNaming.test_ctor_error_handles_ctor_not_opening_name_scope
[ RUN      ] TestModuleNaming.test_does_not_evaluate_property_methods
INFO:tensorflow:time(__main__.TestModuleNaming.test_does_not_evaluate_property_methods): 0.0s
I0301 10:18:35.143716 140670264665920 test_util.py:2082] time(__main__.TestModuleNaming.test_does_not_evaluate_property_methods): 0.0s
[       OK ] TestModuleNaming.test_does_not_evaluate_property_methods
[ RUN      ] TestModuleNaming.test_enters_name_scope_in_call
INFO:tensorflow:time(__main__.TestModuleNaming.test_enters_name_scope_in_call): 0.0s
I0301 10:18:35.144006 140670264665920 test_util.py:2082] time(__main__.TestModuleNaming.test_enters_name_scope_in_call): 0.0s
[       OK ] TestModuleNaming.test_enters_name_scope_in_call
[ RUN      ] TestModuleNaming.test_enters_name_scope_in_other_method
INFO:tensorflow:time(__main__.TestModuleNaming.test_enters_name_scope_in_other_method): 0.0s
I0301 10:18:35.144288 140670264665920 test_util.py:2082] time(__main__.TestModuleNaming.test_enters_name_scope_in_other_method): 0.0s
[       OK ] TestModuleNaming.test_enters_name_scope_in_other_method
[ RUN      ] TestModuleNaming.test_forward_method_closes_name_scope
INFO:tensorflow:time(__main__.TestModuleNaming.test_forward_method_closes_name_scope): 0.0s
I0301 10:18:35.144543 140670264665920 test_util.py:2082] time(__main__.TestModuleNaming.test_forward_method_closes_name_scope): 0.0s
[       OK ] TestModuleNaming.test_forward_method_closes_name_scope
[ RUN      ] TestModuleNaming.test_get_attr_doesnt_enter_name_scope
INFO:tensorflow:time(__main__.TestModuleNaming.test_get_attr_doesnt_enter_name_scope): 0.0s
I0301 10:18:35.144858 140670264665920 test_util.py:2082] time(__main__.TestModuleNaming.test_get_attr_doesnt_enter_name_scope): 0.0s
[       OK ] TestModuleNaming.test_get_attr_doesnt_enter_name_scope
[ RUN      ] TestModuleNaming.test_get_attribute_doesnt_enter_name_scope
INFO:tensorflow:time(__main__.TestModuleNaming.test_get_attribute_doesnt_enter_name_scope): 0.0s
I0301 10:18:35.145172 140670264665920 test_util.py:2082] time(__main__.TestModuleNaming.test_get_attribute_doesnt_enter_name_scope): 0.0s
[       OK ] TestModuleNaming.test_get_attribute_doesnt_enter_name_scope
[ RUN      ] TestModuleNaming.test_invalid_name
INFO:tensorflow:time(__main__.TestModuleNaming.test_invalid_name): 0.0s
I0301 10:18:35.145528 140670264665920 test_util.py:2082] time(__main__.TestModuleNaming.test_invalid_name): 0.0s
[       OK ] TestModuleNaming.test_invalid_name
[ RUN      ] TestModuleNaming.test_module_numbering_in_graph
WARNING:tensorflow:From /usr/local/lib/python3.9/contextlib.py:87: TensorFlowTestCase.test_session (from tensorflow.python.framework.test_util) is deprecated and will be removed in a future version.
Instructions for updating:
Use `self.session()` or `self.cached_session()` instead.
W0301 10:18:35.145755 140670264665920 deprecation.py:330] From /usr/local/lib/python3.9/contextlib.py:87: TensorFlowTestCase.test_session (from tensorflow.python.framework.test_util) is deprecated and will be removed in a future version.
Instructions for updating:
Use `self.session()` or `self.cached_session()` instead.
INFO:tensorflow:time(__main__.TestModuleNaming.test_module_numbering_in_graph): 0.01s
I0301 10:18:35.159946 140670264665920 test_util.py:2082] time(__main__.TestModuleNaming.test_module_numbering_in_graph): 0.01s
INFO:tensorflow:time(__main__.TestModuleNaming.test_module_numbering_in_graph): 0.0s
I0301 10:18:35.160279 140670264665920 test_util.py:2082] time(__main__.TestModuleNaming.test_module_numbering_in_graph): 0.0s
[  SKIPPED ] TestModuleNaming.test_module_numbering_in_graph
[ RUN      ] TestModuleNaming.test_modules_not_numbered_in_eager
INFO:tensorflow:time(__main__.TestModuleNaming.test_modules_not_numbered_in_eager): 0.0s
I0301 10:18:35.160827 140670264665920 test_util.py:2082] time(__main__.TestModuleNaming.test_modules_not_numbered_in_eager): 0.0s
INFO:tensorflow:time(__main__.TestModuleNaming.test_modules_not_numbered_in_eager): 0.0s
I0301 10:18:35.162637 140670264665920 test_util.py:2082] time(__main__.TestModuleNaming.test_modules_not_numbered_in_eager): 0.0s
[       OK ] TestModuleNaming.test_modules_not_numbered_in_eager
[ RUN      ] TestModuleNaming.test_overridden_name_scope
INFO:tensorflow:time(__main__.TestModuleNaming.test_overridden_name_scope): 0.0s
I0301 10:18:35.163025 140670264665920 test_util.py:2082] time(__main__.TestModuleNaming.test_overridden_name_scope): 0.0s
[       OK ] TestModuleNaming.test_overridden_name_scope
[ RUN      ] TestModuleNaming.test_patched_callable
INFO:tensorflow:time(__main__.TestModuleNaming.test_patched_callable): 0.0s
I0301 10:18:35.163293 140670264665920 test_util.py:2082] time(__main__.TestModuleNaming.test_patched_callable): 0.0s
[       OK ] TestModuleNaming.test_patched_callable
[ RUN      ] TestModuleNaming.test_property
INFO:tensorflow:time(__main__.TestModuleNaming.test_property): 0.0s
I0301 10:18:35.163604 140670264665920 test_util.py:2082] time(__main__.TestModuleNaming.test_property): 0.0s
[       OK ] TestModuleNaming.test_property
[ RUN      ] TestModuleNaming.test_property_no_name_scope
INFO:tensorflow:time(__main__.TestModuleNaming.test_property_no_name_scope): 0.0s
I0301 10:18:35.163878 140670264665920 test_util.py:2082] time(__main__.TestModuleNaming.test_property_no_name_scope): 0.0s
[       OK ] TestModuleNaming.test_property_no_name_scope
[ RUN      ] TestModuleNaming.test_session
[  SKIPPED ] TestModuleNaming.test_session
[ RUN      ] TestModuleNaming.test_single_name
INFO:tensorflow:time(__main__.TestModuleNaming.test_single_name): 0.0s
I0301 10:18:35.164222 140670264665920 test_util.py:2082] time(__main__.TestModuleNaming.test_single_name): 0.0s
[       OK ] TestModuleNaming.test_single_name
[ RUN      ] TestModuleNaming.test_subclassed_module
INFO:tensorflow:time(__main__.TestModuleNaming.test_subclassed_module): 0.0s
I0301 10:18:35.164546 140670264665920 test_util.py:2082] time(__main__.TestModuleNaming.test_subclassed_module): 0.0s
[       OK ] TestModuleNaming.test_subclassed_module
[ RUN      ] TestModuleNaming.test_submodule_created_late
INFO:tensorflow:time(__main__.TestModuleNaming.test_submodule_created_late): 0.0s
I0301 10:18:35.164953 140670264665920 test_util.py:2082] time(__main__.TestModuleNaming.test_submodule_created_late): 0.0s
[       OK ] TestModuleNaming.test_submodule_created_late
[ RUN      ] VariableNamingTest.test_session
[  SKIPPED ] VariableNamingTest.test_session
[ RUN      ] VariableNamingTest.test_variable_names
INFO:tensorflow:time(__main__.VariableNamingTest.test_variable_names): 0.0s
I0301 10:18:35.166484 140670264665920 test_util.py:2082] time(__main__.VariableNamingTest.test_variable_names): 0.0s
[       OK ] VariableNamingTest.test_variable_names
[ RUN      ] VariableTrackingTest.test_composite_variable
INFO:tensorflow:time(__main__.VariableTrackingTest.test_composite_variable): 0.0s
I0301 10:18:35.168788 140670264665920 test_util.py:2082] time(__main__.VariableTrackingTest.test_composite_variable): 0.0s
[       OK ] VariableTrackingTest.test_composite_variable
[ RUN      ] VariableTrackingTest.test_session
[  SKIPPED ] VariableTrackingTest.test_session
[ RUN      ] VariableTrackingTest.test_supports_distributed_variables
INFO:tensorflow:time(__main__.VariableTrackingTest.test_supports_distributed_variables): 0.0s
I0301 10:18:35.171150 140670264665920 test_util.py:2082] time(__main__.VariableTrackingTest.test_supports_distributed_variables): 0.0s
[       OK ] VariableTrackingTest.test_supports_distributed_variables
[ RUN      ] VariableTrackingTest.test_trainable_variables
INFO:tensorflow:time(__main__.VariableTrackingTest.test_trainable_variables): 0.0s
I0301 10:18:35.173183 140670264665920 test_util.py:2082] time(__main__.VariableTrackingTest.test_trainable_variables): 0.0s
[       OK ] VariableTrackingTest.test_trainable_variables
[ RUN      ] VariableTrackingTest.test_trainable_variables_ignores_non_trainable
INFO:tensorflow:time(__main__.VariableTrackingTest.test_trainable_variables_ignores_non_trainable): 0.0s
I0301 10:18:35.174891 140670264665920 test_util.py:2082] time(__main__.VariableTrackingTest.test_trainable_variables_ignores_non_trainable): 0.0s
[       OK ] VariableTrackingTest.test_trainable_variables_ignores_non_trainable
[ RUN      ] VariableTrackingTest.test_variables
INFO:tensorflow:time(__main__.VariableTrackingTest.test_variables): 0.0s
I0301 10:18:35.176566 140670264665920 test_util.py:2082] time(__main__.VariableTrackingTest.test_variables): 0.0s
[       OK ] VariableTrackingTest.test_variables
======================================================================
FAIL: testAbstract (__main__.AbcTest)
testAbstract (__main__.AbcTest)
----------------------------------------------------------------------
TypeError: Can't instantiate abstract class AbstractModule with abstract method __call__

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/buildfarm/default/operations/52eafc82-3509-47a0-ad8a-d565ef4de86b/bazel-out/k8-opt/bin/tensorflow/python/module/module_test.runfiles/org_tensorflow/tensorflow/python/module/module_test.py", line 340, in testAbstract
    AbstractModule()  # pylint: disable=abstract-class-instantiated
AssertionError: "Can't instantiate .* abstract methods" does not match "Can't instantiate abstract class AbstractModule with abstract method __call__"

----------------------------------------------------------------------
Ran 50 tests in 0.121s

FAILED (failures=1, skipped=10)
```

**Comment by mdanatg** [comment]
That one looks like a bad test - the check for the error message is too specific and breaks when the error message changes. If anyone is interested in sending t a patch, the fix is straightforward - just update the regex at line 339.

**Comment by park-junha** [comment]
@mdanatg I can open a PR for it

**Comment by park-junha** [comment]
https://github.com/tensorflow/tensorflow/pull/47512 @byronyi @mdanatg should fix the test failure

**Comment by byronyi** [comment]
```
==================== Test output for //tensorflow/tools/api/tests:api_compatibility_test:
Running tests under Python 3.9.2: /usr/local/bin/python3.9
[ RUN      ] ApiCompatibilityTest.testAPIBackwardsCompatibility
INFO:tensorflow:No differences found between API and golden.
I0303 01:00:03.331312 140083892332352 api_compatibility_test.py:321] No differences found between API and golden.
INFO:tensorflow:time(__main__.ApiCompatibilityTest.testAPIBackwardsCompatibility): 2.64s
I0303 01:00:03.337718 140083892332352 test_util.py:2082] time(__main__.ApiCompatibilityTest.testAPIBackwardsCompatibility): 2.64s
[       OK ] ApiCompatibilityTest.testAPIBackwardsCompatibility
[ RUN      ] ApiCompatibilityTest.testAPIBackwardsCompatibilityV1

ERROR:tensorflow:TensorFlow API backwards compatibility test
This test ensures all changes to the public API of TensorFlow are intended.

If this test fails, it means a change has been made to the public API. Backwards
incompatible changes are not allowed. You can run the test as follows to update
test goldens and package them with your change.

    $ bazel run tensorflow/tools/api/tests:api_compatibility_test \
    #     -- --update_goldens True

You will need an API approval to make changes to the public TensorFlow API. This
includes additions to the API.

E0303 01:00:06.100362 140083892332352 api_compatibility_test.py:293] TensorFlow API backwards compatibility test
This test ensures all changes to the public API of TensorFlow are intended.

If this test fails, it means a change has been made to the public API. Backwards
incompatible changes are not allowed. You can run the test as follows to update
test goldens and package them with your change.

    $ bazel run tensorflow/tools/api/tests:api_compatibility_test \
    #     -- --update_goldens True

You will need an API approval to make changes to the public TensorFlow API. This
includes additions to the API.

ERROR:tensorflow:1 differences found between API and golden.
E0303 01:00:06.100567 140083892332352 api_compatibility_test.py:294] 1 differences found between API and golden.
ERROR:tensorflow:    Change detected in python object: tensorflow.train.LooperThread.
E0303 01:00:06.100632 140083892332352 api_compatibility_test.py:315]     Change detected in python object: tensorflow.train.LooperThread.
ERROR:tensorflow:    
  path: "tensorflow.train.LooperThread"
  tf_class {
    is_instance: "<class \'tensorflow.python.training.coordinator.LooperThread\'>"
    is_instance: "<class \'threading.Thread\'>"
    member {
      name: "daemon"
      mtype: "<type \'property\'>"
    }
    member {
      name: "ident"
      mtype: "<type \'property\'>"
    }
    member {
      name: "name"
      mtype: "<type \'property\'>"
    }
    member_method {
      name: "__init__"
      argspec: "args=[\'self\', \'coord\', \'timer_interval_secs\', \'target\', \'args\', \'kwargs\'], varargs=None, keywords=None, defaults=[\'None\', \'None\', \'None\'], "
    }
    member_method {
      name: "getName"
-     argspec: "args=[\'self\'], varargs=None, keywords=None, defaults=None"
-   }
-   member_method {
-     name: "isAlive"
      argspec: "args=[\'self\'], varargs=None, keywords=None, defaults=None"
    }
    member_method {
      name: "isDaemon"
      argspec: "args=[\'self\'], varargs=None, keywords=None, defaults=None"
    }
    member_method {
      name: "is_alive"
      argspec: "args=[\'self\'], varargs=None, keywords=None, defaults=None"
    }
    member_method {
      name: "loop"
      argspec: "args=[\'coord\', \'timer_interval_secs\', \'target\', \'args\', \'kwargs\'], varargs=None, keywords=None, defaults=[\'None\', \'None\'], "
    }
    member_method {
      name: "run"
      argspec: "args=[\'self\'], varargs=None, keywords=None, defaults=None"
    }
    member_method {
      name: "run_loop"
      argspec: "args=[\'self\'], varargs=None, keywords=None, defaults=None"
    }
    member_method {
      name: "setDaemon"
      argspec: "args=[\'self\', \'daemonic\'], varargs=None, keywords=None, defaults=None"
    }
    member_method {
      name: "setName"
      argspec: "args=[\'self\', \'name\'], varargs=None, keywords=None, defaults=None"
    }
    member_method {
      name: "start"
      argspec: "args=[\'self\'], varargs=None, keywords=None, defaults=None"
    }
    member_method {
      name: "start_loop"
      argspec: "args=[\'self\'], varargs=None, keywords=None, defaults=None"
    }
    member_method {
      name: "stop_loop"
      argspec: "args=[\'self\'], varargs=None, keywords=None, defaults=None"
    }
  }

E0303 01:00:06.100685 140083892332352 api_compatibility_test.py:316]     
  path: "tensorflow.train.LooperThread"
  tf_class {
    is_instance: "<class \'tensorflow.python.training.coordinator.LooperThread\'>"
    is_instance: "<class \'threading.Thread\'>"
    member {
      name: "daemon"
      mtype: "<type \'property\'>"
    }
    member {
      name: "ident"
      mtype: "<type \'property\'>"
    }
    member {
      name: "name"
      mtype: "<type \'property\'>"
    }
    member_method {
      name: "__init__"
      argspec: "args=[\'self\', \'coord\', \'timer_interval_secs\', \'target\', \'args\', \'kwargs\'], varargs=None, keywords=None, defaults=[\'None\', \'None\', \'None\'], "
    }
    member_method {
      name: "getName"
-     argspec: "args=[\'self\'], varargs=None, keywords=None, defaults=None"
-   }
-   member_method {
-     name: "isAlive"
      argspec: "args=[\'self\'], varargs=None, keywords=None, defaults=None"
    }
    member_method {
      name: "isDaemon"
      argspec: "args=[\'self\'], varargs=None, keywords=None, defaults=None"
    }
    member_method {
      name: "is_alive"
      argspec: "args=[\'self\'], varargs=None, keywords=None, defaults=None"
    }
    member_method {
      name: "loop"
      argspec: "args=[\'coord\', \'timer_interval_secs\', \'target\', \'args\', \'kwargs\'], varargs=None, keywords=None, defaults=[\'None\', \'None\'], "
    }
    member_method {
      name: "run"
      argspec: "args=[\'self\'], varargs=None, keywords=None, defaults=None"
    }
    member_method {
      name: "run_loop"
      argspec: "args=[\'self\'], varargs=None, keywords=None, defaults=None"
    }
    member_method {
      name: "setDaemon"
      argspec: "args=[\'self\', \'daemonic\'], varargs=None, keywords=None, defaults=None"
    }
    member_method {
      name: "setName"
      argspec: "args=[\'self\', \'name\'], varargs=None, keywords=None, defaults=None"
    }
    member_method {
      name: "start"
      argspec: "args=[\'self\'], varargs=None, keywords=None, defaults=None"
    }
    member_method {
      name: "start_loop"
      argspec: "args=[\'self\'], varargs=None, keywords=None, defaults=None"
    }
    member_method {
      name: "stop_loop"
      argspec: "args=[\'self\'], varargs=None, keywords=None, defaults=None"
    }
  }

INFO:tensorflow:time(__main__.ApiCompatibilityTest.testAPIBackwardsCompatibilityV1): 2.76s
I0303 01:00:06.100801 140083892332352 test_util.py:2082] time(__main__.ApiCompatibilityTest.testAPIBackwardsCompatibilityV1): 2.76s
[  FAILED  ] ApiCompatibilityTest.testAPIBackwardsCompatibilityV1
[ RUN      ] ApiCompatibilityTest.testAPIBackwardsCompatibilityV2
INFO:tensorflow:No differences found between API and golden.
I0303 01:00:08.738246 140083892332352 api_compatibility_test.py:321] No differences found between API and golden.
INFO:tensorflow:time(__main__.ApiCompatibilityTest.testAPIBackwardsCompatibilityV2): 2.64s
I0303 01:00:08.744600 140083892332352 test_util.py:2082] time(__main__.ApiCompatibilityTest.testAPIBackwardsCompatibilityV2): 2.64s
[       OK ] ApiCompatibilityTest.testAPIBackwardsCompatibilityV2
[ RUN      ] ApiCompatibilityTest.testNoSubclassOfMessage
INFO:tensorflow:time(__main__.ApiCompatibilityTest.testNoSubclassOfMessage): 0.34s
I0303 01:00:09.094376 140083892332352 test_util.py:2082] time(__main__.ApiCompatibilityTest.testNoSubclassOfMessage): 0.34s
[       OK ] ApiCompatibilityTest.testNoSubclassOfMessage
[ RUN      ] ApiCompatibilityTest.testNoSubclassOfMessageV1
INFO:tensorflow:time(__main__.ApiCompatibilityTest.testNoSubclassOfMessageV1): 0.34s
I0303 01:00:09.430554 140083892332352 test_util.py:2082] time(__main__.ApiCompatibilityTest.testNoSubclassOfMessageV1): 0.34s
[       OK ] ApiCompatibilityTest.testNoSubclassOfMessageV1
[ RUN      ] ApiCompatibilityTest.testNoSubclassOfMessageV2
INFO:tensorflow:time(__main__.ApiCompatibilityTest.testNoSubclassOfMessageV2): 0.32s
I0303 01:00:09.751907 140083892332352 test_util.py:2082] time(__main__.ApiCompatibilityTest.testNoSubclassOfMessageV2): 0.32s
[       OK ] ApiCompatibilityTest.testNoSubclassOfMessageV2
[ RUN      ] ApiCompatibilityTest.test_session
[  SKIPPED ] ApiCompatibilityTest.test_session
======================================================================
FAIL: testAPIBackwardsCompatibilityV1 (__main__.ApiCompatibilityTest)
testAPIBackwardsCompatibilityV1 (__main__.ApiCompatibilityTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/buildfarm/default/operations/7d0cf0f7-4e69-4535-a932-0a1eee6aad4f/bazel-out/k8-opt/bin/tensorflow/tools/api/tests/api_compatibility_test.runfiles/org_tensorflow/tensorflow/tools/api/tests/api_compatibility_test.py", line 440, in testAPIBackwardsCompatibilityV1
    self._checkBackwardsCompatibility(
  File "/buildfarm/default/operations/7d0cf0f7-4e69-4535-a932-0a1eee6aad4f/bazel-out/k8-opt/bin/tensorflow/tools/api/tests/api_compatibility_test.runfiles/org_tensorflow/tensorflow/tools/api/tests/api_compatibility_test.py", line 400, in _checkBackwardsCompatibility
    self._AssertProtoDictEquals(
  File "/buildfarm/default/operations/7d0cf0f7-4e69-4535-a932-0a1eee6aad4f/bazel-out/k8-opt/bin/tensorflow/tools/api/tests/api_compatibility_test.runfiles/org_tensorflow/tensorflow/tools/api/tests/api_compatibility_test.py", line 318, in _AssertProtoDictEquals
    self.fail('%d differences found between API and golden.' % diff_count)
  File "/buildfarm/default/operations/7d0cf0f7-4e69-4535-a932-0a1eee6aad4f/bazel-out/k8-opt/bin/tensorflow/tools/api/tests/api_compatibility_test.runfiles/absl_py/absl/testing/absltest.py", line 1704, in fail
    return super(TestCase, self).fail(self._formatMessage(prefix, msg))
AssertionError: 1 differences found between API and golden.

----------------------------------------------------------------------
Ran 7 tests in 9.058s

FAILED (failures=1, skipped=1)
```

I ran Python 3.6-3.8 tests prior to Python 3.9 and all passed.

/cc @mdanatg

**Comment by byronyi** [comment]
It's a v1 test so may not be that important. Context: https://bugs.python.org/issue37804

**Comment by mdanatg** [comment]
Makes sene. Looks like that test is brittle, too. It seems that the symbol is already special-cased by the test: https://github.com/tensorflow/tensorflow/blob/27cb1e3e60d03be9262351795f46690ada5797e3/tensorflow/tools/api/lib/python_object_to_proto_visitor.py#L48

Perhaps just adding isAlive in there will be sufficient.

**Comment by ShyamSundhar1411** [comment]
Any updates on this issue? I am unable to install Tensorflow in my windows py 3.9

**Comment by bhack** [comment]
> Any updates on this issue? I am unable to install Tensorflow in my windows py 3.9

Have you tried with tf-nightly?

**Comment by mihaimaruseac** [comment]
TF nightly now has support for python 3.9 on all 3 operating systems.

Please test it and report failures. We'll try to get them fixed before the next release.

Keeping this open until next release. Branch cut is in 2 weeks and maybe 4 more weeks of RCs after that, so likely near the end of April.

**Comment by bhack** [comment]
@mihaimaruseac As It is quite popular fan we pin this issue in the meantime? What do you think?

**Comment by cclauss** [comment]
Great news!  Can we start testing on Python 3.10.0 alpha 6?  It would be good to have a shorter ramp up next time.

**Comment by mihaimaruseac** [comment]
The largest part of the ramp up is because of upstream dependencies. TF cannot build until they can build with the new version of Python. In particular, numpy, protobuf, grpc are the main delayers.

**Comment by ShyamSundhar1411** [comment]
But tf nightly is not getting imported . I am getting ModuleNotFoundError

**Comment by mihaimaruseac** [comment]
@ShyamSundhar1411 this looks like a different issue. Make sure you use the same python as the one used to do `pip install`. Recommendation is to use `python -m pip install` instead of `pip install`.

If the above does not fix, please open a new issue and fill in issue template and give a full description of the error you are seeing. Also, give output of both `python -m pip install`

**Comment by cclauss** [comment]
_Explicit is better than implicit..._`python3.9 -m pip install xyz`

**Comment by staticdev** [comment]
@mihaimaruseac this is marked for TF 2.5 but there is no mention of it on the current [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md).

**Comment by mihaimaruseac** [comment]
RC0 [files](https://pypi.org/project/tensorflow/2.5.0rc0/#files) already contain Py3.9 wheels.

We'll update the release notes, it seems we don't always mention support for new versions of Python in there.

**Comment by cclauss** [comment]
I think this is fixed now.

**Comment by Nizarazo** [comment]
Is Keras module that exists in tensorflow supported in python 3.9?

**Comment by felixdivo** [comment]
@mihaimaruseac Seems like this issue can be closed.
